src/main/java/eu/dissco/core/digitalspecimenprocessor/service/MidsService.java

```diff
@@ -54,88 +57,175 @@ private boolean compliesToMidsTwo(DigitalSpecimen attributes) {
     }
   }
 
+  private boolean compliesToMidsThree(DigitalSpecimen attributes) {
+    var topComplies = hasValue(attributes.getDwcInstitutionId());
+    var taxValuesComply = taxValuesComply(attributes);
+    var geographicalValuesComply = geographicalValuesComply(attributes);
+    if (PALEO_GEO_DISCIPLINES.contains(attributes.getOdsTopicDiscipline())) {
+      return topComplies && taxValuesComply && geographicalValuesComply;
+    } else {
+      return topComplies && taxValuesComply && geographicalValuesComply && isValid(
+          attributes.getDwcRecordedById());
+    }
+  }
+
+  private boolean geographicalValuesComply(DigitalSpecimen attributes) {
+    if (attributes.getDwcOccurrence() != null
+        && attributes.getDwcOccurrence().get(0).getDctermsLocation() != null) {
+      var location = attributes.getDwcOccurrence().get(0).getDctermsLocation();
+      if (isValid(location.getDwcLocationID())) {
+        return true;
+      } else if (location.getGeoReference() != null) {
+        var georeference = location.getGeoReference();
+        return (georeference.getDwcDecimalLatitude() != null
+            && georeference.getDwcDecimalLongitude() != null
+            && isValid(georeference.getDwcGeodeticDatum())
+            && georeference.getDwcCoordinateUncertaintyInMeters() != null
+            && georeference.getDwcCoordinatePrecision() != null)
+            || (isValid(georeference.getDwcFootprintWkt())
+            && isValid(georeference.getDwcFootprintSrs()));
+      }
+    }
+    return false;
+  }
+
+  private boolean taxValuesComply(DigitalSpecimen attributes) {
+    var hasScientificNameId = false;
+    var hasIdentifiedById = false;
+    for (var identifications : attributes.getDwcIdentification()) {
+      if (isValid(identifications.getDwcIdentificationID())) {
+        hasIdentifiedById = true;
+      }
```

Comment:
could already return false if hasIdentifiedById is false, that way you don't cycle through taxonIdentifications

Reply:
But that can only be when that is the last hasIdentifierById is false. Let's say there are 5 identifications, the first one has a scientificNameId and the last a IdentificationID. I am now assuming that is fine, somewhere in the document these fields are filled (does not matter that it is not within a single identification). I made this double loop so we don't have to go through all identification twice, but can just do one run for both fields. Not sure if I explain it well enough. Did pick the wrong field for the evaluation though :smiley_cat: (so this comment still saved us a bug :bug: )

Reply:
oops i just missed the looping! happy to inadvertently help tho